wa/instrumentation/hwmon.py

```diff
@@ -0,0 +1,88 @@
+#    Copyright 2017 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from devlib import HwmonInstrument as _Instrument
+
+from wa import Instrument
+from wa.framework.instrumentation import fast
+
+MOMENTARY_QUANTITIES = ['temperature', 'power', 'voltage', 'current', 'fps']
+CUMULATIVE_QUANTITIES = ['energy', 'tx', 'tx/rx', 'frames']
+
+class HwmonInstrument(Instrument):
+    name = 'hwmon'
+
+    description = """
+    Hardware Monitor (hwmon) is a generic Linux kernel subsystem,
+    providing access to hardware monitoring components like temperature or
+    voltage/current sensors.
+
+    Data from hwmon that are a snapshot of a fluctuating value, such as
+    temperature and voltage, are reported once at the beginning and once at the
+    end of the workload run. Data that are a cumulative total of a quantity,
+    such as energy (which is the cumulative total of power consumption), are
+    reported as the difference between the values at the beginning and at the
+    end of the workload run.
+
+    There is currently no functionality to filter sensors: all of the available
+    hwmon data will be reported.
+    """
+
+    def initialize(self, context):
+        self.instrument = _Instrument(self.target)
+
+    def setup(self, context):
+        self.instrument.reset()
+
+    @fast
```

Comment:
Since Energy measurements start/stop should be the really first thing we start/stop across the execution of a workload... should not be safer to set these callbacks as @very_fast level?

Reply:
"@fast" already means this will run closer to the measured region than most other instrumentation. Bear in mind, that the decorator is called that for a reason -- it implies that the method it decorates is actually fast to execute and will not interfere much with the instrumentation that is already running. Given that take_measurement() for hwmon will result in a call to the target for each active channel, this definitely does not qualify as "@very_fast" (if anything, I would argue that "@fast" is a stretch...).